### PR TITLE
Refactor Lists to Arrays

### PR DESCRIPTION
**Pull request by JonoPrest**
- Remove use of lists holding eventBatchQueueItem
- Where nodes of lists were returned, replace with a lazy computation on the original orray

### CODE REVIEW

**Comment by JonoPrest** [comment]
Busy fixing a few bugs 👍🏼

**Comment by JonoPrest** [comment]
Hey @DZakh, the organize utils PR is based on top of this so I won't merge it until this is reviewed so it's a bit easier to separate out the changes.

File: codegenerator/cli/templates/static/codegen/src/EventProcessing.res [hunk 1]
```diff
@@ -284,24 +282,24 @@ let addToUnprocessedBatch = (
...
   ~dynamicContractRegistrations: option<dynamicContractRegistrations>=None,
   ~inMemoryStore,
-  eventBatch: list<Types.eventBatchQueueItem>,
 ) => {
-  switch eventBatch {
-  | list{} => (eventsBeforeDynamicRegistrations, dynamicContractRegistrations)->Ok
-  | list{eventBatchQueueItem, ...tail} =>
+  switch eventBatch[index] {
+  | None => (eventsBeforeDynamicRegistrations, dynamicContractRegistrations)->Ok
+  // | list{eventBatchQueueItem, ...tail} =>
```
Comment (JonoPrest) [inline codegenerator/cli/templates/static/codegen/src/EventProcessing.res:301]:
```suggestion
```

File: codegenerator/cli/templates/static/codegen/src/eventFetching/ChainManager.res [hunk 2]
```diff
@@ -193,40 +187,42 @@ let setChainFetcher = (self: t, chainFetcher: ChainFetcher.t) => {
...
 type earliestQueueItem =
-  | ArbitraryEventQueue(Types.eventBatchQueueItem, list<Types.eventBatchQueueItem>)
+  | ArbitraryEventQueue(Types.eventBatchQueueItem, array<Types.eventBatchQueueItem>)
   | EventFetchers(Types.eventBatchQueueItem, ChainMap.t<PartitionedFetchState.t>)
 
 let rec getFirstArbitraryEventsItemForChain = (
-  ~revHead=list{},
+  queue: array<Types.eventBatchQueueItem>,
+  ~index=0,
+  ~head=[],
```
Comment (JonoPrest) [inline codegenerator/cli/templates/static/codegen/src/eventFetching/ChainManager.res:196]: Noticed I don't need this anymore
```suggestion
```

File: codegenerator/cli/templates/static/codegen/src/eventFetching/ChainManager.res [hunk 3]
```diff
@@ -193,40 +187,42 @@ let setChainFetcher = (self: t, chainFetcher: ChainFetcher.t) => {
...
+  | None => None
+  | Some(first) =>
+    let nextIndex = index + 1
     if first.chain == chain {
-      let rest = revHead->List.reverseConcat(tail)
-      Some((first, rest))
+      Some((first, () => queue->Utils.Array.removeAtIndex(index)))
     } else {
-      tail->getFirstArbitraryEventsItemForChain(~chain, ~revHead=list{first, ...revHead})
+      let _ = head->Js.Array2.push(first)
```
Comment (JonoPrest) [inline codegenerator/cli/templates/static/codegen/src/eventFetching/ChainManager.res:206]:
```suggestion                             
```

**Review by DZakh — COMMENTED** [review]
I think ideally for the arbitaryQueue instead of merging everything into a single array with sorting and cloning it on every change we should have:

1. An array for each chain (and maybe for every contract registration)
2. Metadata which item we consumed for every array
3. Metadata which array is consuming and a condition when it should switch to processing another one (we can get it from the smallest eventCmp of the first array item)

- This way it's easier to find an item to process for a specific chain
- Don't need to apply sorting for every item, and only run the check to figure out from which array to get the item next
- Don't need to recreate an array on consuming an item - only increase the pointer index
- When adding more items, just concat them to the array

But it's a bigger refactoring, I don't say we need to do it now. I'll continue reviewing the current changes.

File: codegenerator/cli/templates/static/codegen/src/eventFetching/ChainManager.res [hunk 4]
```diff
@@ -306,13 +297,12 @@ let rec createBatchInternal = (
         )
       | EventFetchers(item, fetchStatesMap) => (arbitraryEventQueue, fetchStatesMap, item)
       }
+      let _ = batch->Js.Array2.push(nextItem)
```
Comment (DZakh) [inline codegenerator/cli/templates/static/codegen/src/eventFetching/ChainManager.res:300]: Is it ok to push here?

**Review by DZakh — APPROVED** [review]
This looks amazing 🚀

Still a little bit painful to look at all the slices and concats, but already a huge improvement!

File: codegenerator/cli/templates/static/codegen/src/eventFetching/ChainManager.res (hunk 4 above)
Comment (JonoPrest) [reply codegenerator/cli/templates/static/codegen/src/eventFetching/ChainManager.res:300]: Yeah "batch" is internal to this function.

**Comment by JonoPrest** [comment]
Cool. Lets chat about your points in person, it will be easier 👍🏼 some of them very valid but as you said part of a bigger refactor.

**Pull request merged**